File: python_Factory_design_pattern/car_factory.py
__author__ = 'Aseem Jain'

class Car(object):

    def factory(type):
        if type == "Racecar":
            return Racecar()
        if type == "Van":
            return Van()
        if type == "SUV":
            return Van()
        assert 0, "Bad car creation: " + type

    factory = staticmethod(factory)

class Racecar(Car):
    def drive(self):
        print("Racecar driving.")


class Van(Car):
    def drive(self):
        print("Van driving.")

class SUV(Car):
    def drive(self):
        print("SUV driving.")

# Create object using factory.
print("Car factory creating Cars based on params ")
car = Car.factory("Racecar")
car.drive()

# Create object using factory.
car = Car.factory("Van")
car.drive()

File: carFactory/src/com/premaseem/CarFactory.java
package com.premaseem;

/*
@author: Aseem Jain
@title: Design Patterns with Java 9
@link: https://premaseem.wordpress.com/category/computers/design-patterns/
@copyright: 2018 Packt Publication
*/

//Class: Factory Design pattern to generate cars based on input parameter
public class CarFactory {

    public static void main(String[] args) {
        System.out.println("Car factory creating Cars based on params ");
        Generator generator = new Generator();

        Car car = generator.factory("racecar");
        car.drive();

        car = generator.factory("Van");
        car.drive();
    }

}

class Generator{
    Car factory(String type){
        if (type.equalsIgnoreCase("racecar")){
            return new Racecar();
        }
        if (type.equalsIgnoreCase("Van")){
            return new Van();
        }
        if (type.equalsIgnoreCase("SUV")){
            return new SUV();
        }
        return null;
    }
}
interface Car{
    void drive();
}

class Racecar implements Car{
    public void drive(){
        System.out.println("Racecar driving.");
    }
}

class Van implements Car{
    public void drive(){
        System.out.println("Van driving.");
    }
}

class SUV implements Car{
    public void drive(){
        System.out.println("SUV driving.");
    }
}


